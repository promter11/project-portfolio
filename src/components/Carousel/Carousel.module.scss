@import "../../styles/_variables";
@import "../../styles/_mixins";

.swiper {
  @include flex(nowrap, column-reverse, flex-start, flex-start);

  .slide {
    height: auto;
  }

  .wrapper {
    @include flex(nowrap, row, space-between, center);
    width: 100%;
    margin-bottom: 4rem;

    .inner {
      .count {
        @include text(4.8rem, 900, 1, normal);
        color: $white-color;
      }

      .button {
        @include text(2.4rem, 900, 150%, uppercase);
        color: $white-color;

        &:not(:last-child) {
          margin-right: 4rem;
        }
      }
    }
  }
}
