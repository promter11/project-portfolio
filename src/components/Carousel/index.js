import React, { useRef } from "react";
import { string, arrayOf, shape } from "prop-types";
import SwiperCore, { Navigation, Pagination } from "swiper";
import { Swiper, SwiperSlide } from "swiper/react";

import Slide from "../Slide";

import Prev from "../../assets/icons/portfolio/arrow-left.svg";
import Next from "../../assets/icons/portfolio/arrow-right.svg";
import "swiper/swiper.scss";
import styles from "./Carousel.module.scss";

SwiperCore.use([Navigation, Pagination]);

const Carousel = ({ slides }) => {
  const prevRef = useRef(null);
  const nextRef = useRef(null);
  const paginationRef = useRef(null);

  return (
    <Swiper
      className={styles.swiper}
      spaceBetween={40}
      navigation={{
        prevEl: prevRef.current,
        nextEl: nextRef.current,
      }}
      pagination={{
        el: paginationRef.current,
        clickable: true,
        type: "fraction",
      }}
      breakpoints={{
        0: {
          slidesPerView: 1,
        },
        440: {
          slidesPerView: 1.6,
        },
        768: {
          slidesPerView: 2.4,
        },
        1920: {
          slidesPerView: 3,
        },
      }}
      onInit={(swiper) => {
        swiper.params.navigation.prevEl = prevRef.current;
        swiper.params.navigation.nextEl = nextRef.current;
        swiper.params.pagination.el = paginationRef.current;
      }}
    >
      <div className={styles.wrapper}>
        <div className={styles.inner}>
          <div className={styles.count} ref={paginationRef} />
        </div>
        <div className={styles.inner}>
          <button className={styles.button} ref={prevRef}>
            <img className={styles.icon} src={Prev} alt="Prev" />
          </button>
          <button className={styles.button} ref={nextRef}>
            <img className={styles.icon} src={Next} alt="Next" />
          </button>
        </div>
      </div>
      {slides.map((slide, index) => {
        return (
          <SwiperSlide className={styles.slide} key={index}>
            <Slide {...slide} />
          </SwiperSlide>
        );
      })}
    </Swiper>
  );
};

Carousel.propTypes = {
  slides: arrayOf(
    shape({
      title: string,
      date: string,
      description: string,
      fullDescription: string,
      technologies: arrayOf(string),
      image: string,
      link: shape({
        text: string,
        path: string,
      }),
    })
  ).isRequired,
};

export default Carousel;
