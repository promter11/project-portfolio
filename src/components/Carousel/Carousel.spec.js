import React from "react";
import { shallow } from "enzyme";

import Carousel from "./index";

const props = {
  slides: [
    {
      title: "Title",
      date: "December 24, 2000",
      description: "Short description",
      fullDescription: "Full description",
      technologies: ["1", "2"],
      image: "../images/1",
      link: {
        text: "Learn more",
        path: "/",
      },
    },
    {
      title: "Title",
      date: "October 12, 1995",
      description: "Short description",
      fullDescription: "Full description",
      technologies: ["1", "2", "3"],
      image: "../images/2",
      link: {
        text: "Learn more",
        path: "/",
      },
    },
  ],
};

describe("Carousel component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Carousel {...props} />);
  });

  it("renders correctly with items", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("check onInit function in Swiper component", () => {
    const mockFn = jest.fn();

    wrapper.find(".swiper").props.onInit = mockFn;
    wrapper.find(".swiper").props.onInit();

    expect(mockFn).toHaveBeenCalled();
  });
});
