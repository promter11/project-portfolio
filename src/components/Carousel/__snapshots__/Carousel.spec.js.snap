// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel component renders correctly with items 1`] = `
<Swiper
  breakpoints={
    Object {
      "0": Object {
        "slidesPerView": 1,
      },
      "1920": Object {
        "slidesPerView": 3,
      },
      "440": Object {
        "slidesPerView": 1.6,
      },
      "768": Object {
        "slidesPerView": 2.4,
      },
    }
  }
  className="swiper"
  navigation={
    Object {
      "nextEl": null,
      "prevEl": null,
    }
  }
  onInit={[Function]}
  pagination={
    Object {
      "clickable": true,
      "el": null,
      "type": "fraction",
    }
  }
  spaceBetween={40}
>
  <div
    className="wrapper"
  >
    <div
      className="inner"
    >
      <div
        className="count"
      />
    </div>
    <div
      className="inner"
    >
      <button
        className="button"
      >
        <img
          alt="Prev"
          className="icon"
          src="arrow-left.svg"
        />
      </button>
      <button
        className="button"
      >
        <img
          alt="Next"
          className="icon"
          src="arrow-right.svg"
        />
      </button>
    </div>
  </div>
  <SwiperSlide
    className="slide"
    key="0"
  >
    <Slide
      date="December 24, 2000"
      description="Short description"
      fullDescription="Full description"
      image="../images/1"
      link={
        Object {
          "path": "/",
          "text": "Learn more",
        }
      }
      technologies={
        Array [
          "1",
          "2",
        ]
      }
      title="Title"
    />
  </SwiperSlide>
  <SwiperSlide
    className="slide"
    key="1"
  >
    <Slide
      date="October 12, 1995"
      description="Short description"
      fullDescription="Full description"
      image="../images/2"
      link={
        Object {
          "path": "/",
          "text": "Learn more",
        }
      }
      technologies={
        Array [
          "1",
          "2",
          "3",
        ]
      }
      title="Title"
    />
  </SwiperSlide>
</Swiper>
`;
